First version of transition notify transformation into shake detect

ariasfelipe opened the pull request.

felipefelixarias committed:
First version of transition notify transformation into shake detect

shankari commented on res/android/statemachine.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
```
You don't need this file, which is the mapping from state names to strings.

shankari commented on src/android/TransitionNotificationReceiver.java:
```diff
@@ -0,0 +1,402 @@
+package edu.berkeley.eecs.emission.cordova.transitionnotify;
```
Why do you have this? I thought you were going with the approach of having the service generate the notification directly.

shankari commented on src/android/shakeNotifier.java:
```diff
@@ -0,0 +1,246 @@
+package edu.berkeley.eecs.emission.cordova.shakeDetect;
```
Does this even work? I thought java required filenames to match classnames, so this would be `ShakeNotifier`.

shankari commented on src/android/shakeNotifier.java:
```diff
@@ -0,0 +1,246 @@
+    private static final String MUTED_LIST_KEY = "muted_list";
+    private static final String ID = "id";
+
+    java.util.Map<String,BroadcastReceiver> receiverMap =
```
don't need reciever map

shankari commented on www/transitionnotify.js:
```diff
@@ -0,0 +1,56 @@
+/*global cordova, module*/
```
remove this file

shankari replied:
file still needs to be removed.

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,222 @@
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId){
+        sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
```
when do you start this service? and when do you stop it?

shankari replied:
Still don't see where service is started/stopped.

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,222 @@
+        if(deltaTime > 500) { //Make the 500 a const
+
+            if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
+                float xAcceleration = event.values[0];
```
where are you storing values into the database?

shankari replied:
if you stored values in the database, you could use a sequence of values to detect the shake instead of a single reading.

shankari replied:
still don't see where values are stored in the database

shankari replied:
note that database != memory, and I routinely store location and motion activity data into the database

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,222 @@
+            }
+            else if(event.sensor.getType() == Sensor.TYPE_GYROSCOPE){
+                float xRotation = event.values[0];
+                String GyroTag = "2Gyro5";
```
Ditto

shankari reviewed (changes requested):
High level changes before I spend more time reviewing.
- Please remove unnecessary files - I don't want to waste time reviewing things that are not necessary

Lots of other specific, but still high level comments...

felipefelixarias committed:
Version 2
VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV

shankari commented on plugin.xml:
```diff
@@ -46,20 +46,18 @@
     
     <config-file target="config.xml" parent="/*">
-      <feature name="TransitionNotification">
+      <feature name="Shake-Detection">
```
why not keep the feature name consistent across android and iOS?

shankari commented on src/android/shakeDetector.java:
```diff
@@ -1,51 +1,54 @@
+        import edu.berkeley.eecs.emission.cordova.unifiedlogger.Log;
+        import edu.berkeley.eecs.emission.cordova.unifiedlogger.NotificationHelper;
+
+public class ShakeDetector extends Service implements SensorEventListener{
```
classname <-> filename mismatch

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,258 @@
+
+
+        //Only get readings if an "accident" has not occured within the past 0.5 seconds
+        if(deltaTime > 500) { //Make the 500 a const
```
actually make it configurable as part of the JSON config that is passed in

shankari commented on src/android/shakeNotifier.java:
```diff
@@ -0,0 +1,245 @@
+    private static final String CONFIG_LIST_KEY = "config_list";
+    private static final String MUTED_LIST_KEY = "muted_list";
+    private static final String ID = "id";
+    private static final String CUSTOM_SHAKE = "custom_shake";
```
where is this even used?

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,258 @@
+                    NotificationHelper.createNotification(this, Constants.TRACKING_ERROR_ID, GyroTag);
+                    //Log.i(this, GyroTag, "," + Float.toString(xRotation) + "," );
+                }*/
+                notifyEvent(this, GyroTag, new JSONObject());
```
if your eventName tags are things like `2Gyro5`, you should probably put them into the javascript so that people can configure listeners for them

shankari commented on www/shakedetect.js:
```diff
@@ -0,0 +1,55 @@
+ */
+
+var TransitionNotification = {
+    INCIDENT: 'incident_detected',
```
these should match the events that can actually be generated from the transition code
doesn't make sense to have `NO_INCIDENT` because that is the default state, not a transition.

shankari commented on www/ui/ionic/js/transition-notify-settings.js:
```diff
@@ -0,0 +1,255 @@
+angular.module('emission.main.control.tnotify', [])
```
file needs to be removed

shankari commented on www/ui/ionic/templates/main-transition-notify-settings.html:
```diff
@@ -0,0 +1,16 @@
+<ion-popover-view style="width: 75%;">
```
file needs to be removed

shankari commented on www/ui/ionic/templates/transition-notify-display-detail.html:
```diff
@@ -0,0 +1,4 @@
+              <div class="col timestamp item-text-wrap control-info">{{entry.transitionName}}</div>
```
file needs to be removed

shankari reviewed (changes requested):
Still have a bunch of obsolete transition filenames
Still need to start/stop the service
Need consistency with event names

felipefelixarias committed:
Version 3

Updated variable names and went through all javascript and html files. Fixed the issues brought up in the comments but almost all of the work was done on the javascript files. Next steps:
Implement broadcastReceiver to initialize service
Store accelerometer readings in database
Debug

shankari commented on www/ui/ionic/js/shake-detect-settings.js:
```diff
@@ -0,0 +1,245 @@
+.factory("ControlShakeDetectHelper", function($window, 
+        $ionicActionSheet, $ionicPopup, $ionicPopover, $rootScope) {
+
+    var ctnh = {};
```
nit: `ctnh` was `ControlTransitionNotifyHelper`

shankari replied:
So yours should really be `csdh`...

shankari reviewed (commented):
javascript changes seem fine.
looking forward to native fixes

felipefelixarias committed:
Created shake notification receiver
Transformed transitionNotificationReceiver into a broadcast reciever to enable and disable the shakeDetector service (Am I doing it at the right transitions?). Updated shake detector to check if the sensors are available in the following order of priority: gyroscope, linear accelerometer, accelerometer. Changed variable names from ctnh to csdh in shake-detect-settings.js.

felipefelixarias committed:
Added shakeNotificationReceiver.java

felipefelixarias committed:
First Working shakedetect Plugin
Updated plugin.xml to include service and receiver. Corrections to javascript files (notification not currently working). Syntax corrections to one of the java files (simple name change to match class name to file name).

felipefelixarias committed:
Second version of working Plugin
Rearranged ShakeDetector.java to use the accelerometer over the gyroscope. Fixed syntax errors and updated other files for notifications to work (still not working).

shankari commented on src/android/shakeDetector.java:
```diff
@@ -5,33 +5,36 @@
 /*
  * Importing dependencies from the logger plugin
  */
+import edu.berkeley.eecs.emission.cordova.tracker.Constants;
```
Don't introduce a dependency from this plugin to `tracker` (DataCollection) unless you need to.
Are you actually using this anywhere? Most usage appears to be commented out...

shankari commented on src/android/shakeDetector.java:
```diff
@@ -5,33 +5,36 @@
  */
+import edu.berkeley.eecs.emission.cordova.tracker.Constants;
 import edu.berkeley.eecs.emission.cordova.unifiedlogger.Log;
+import edu.berkeley.eecs.emission.cordova.unifiedlogger.NotificationHelper;
```
Why are you using NotificationHelper?
You want to use the notification plugin through the `notifyEvent` function

shankari commented on src/android/shakeDetector.java:
```diff
@@ -106,43 +110,27 @@ public void onSensorChanged(SensorEvent event){
-                    notifyEvent(this, AccTag1, new JSONObject());
+                    //NotificationHelper.createNotification(this, Constants.TRACKING_ERROR_ID, AccTag1);
+                    Log.i(this, AccTag1, ","+String.valueOf(currTime)+"," + Float.toString(xAcceleration) + "," + Float.toString(yAcceleration) + "," + Float.toString(zAcceleration) + "," + Float.toString(acceleration) + ",");
+                    notifyEvent(this, POTENTIAL_INCIDENT, new JSONObject());
```
future fix: you could potentially have the `'data'` part of the event represent the sensor that detected the gesture - e.g. {'sensor': 'accelerometer'}

shankari commented on src/android/shakeDetector.java:
```diff
@@ -157,37 +145,52 @@ else if(event.sensor.getType() == Sensor.TYPE_LINEAR_ACCELERATION){
+        Intent shakeDetectionIntent = new Intent();
+        shakeDetectionIntent.setAction(eventName);
+        shakeDetectionIntent.putExtras(jsonToBundle(autogenData));
+        context.sendBroadcast(shakeDetectionIntent);
```
the newly added code is not really notifying an event, it is broadcasting an intent.
you should really refactor into `broadcastAndNotifyEvent` as I think I did.

shankari commented on src/android/shakeDetector.java:
```diff
@@ -248,4 +251,43 @@ private int findEntryWithId(JSONArray array, long id) throws JSONException {
         return -1;
     }
+
+        private Bundle jsonToBundle(JSONObject toConvert) {
```
where do you call this?

shankari reviewed (changes requested):
I think you are almost there.
This *should* work - you just need to fix the phone code as I described in my review there.

shankari commented:
Also, you need to store to the database. I don't see that change here yet.

felipefelixarias committed:
Slight Changes
Made slight changes to ShakeDetector.java (some of the recommendations from the last pull request).
Currently, JSONObject notifyConfigWrapper = UserCacheFactory.getUserCache(context).getLocalStorage(eventName, false); in ShakeDetector.java line 193 returns null which results in no notification.

felipefelixarias committed:
First Version of iOS Plugin
First Version of iOS Plugin (not yet tested).

felipefelixarias committed:
iOS and changes to attempt to make notifications work
Almost done with iOS (have to debug to receive the geo-fence and start trip transitions) and made changes to attempt to make notifications work (no luck).

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,253 @@
+                            mergeObjects(currData, autogenData);
+                        }
+                        Log.d(context, TAG, "generating notification for event "+eventName
+                                + " and id = " + currNotifyConfig.getLong(ID));
```
My suggestion is to clone the `currNotifyConfig` here and make the id of the cloned config = orig id + current time or something unique, and then pass the cloned config to be scheduled.

shankari replied:
This suggestion has not yet been implemented.

felipefelixarias committed:
First Working Version of Android and iOS
Notifications are working on both platforms and iOS code was brought up to date. Bug: on iOS, the notification seems at appear twice.

felipefelixarias committed:
Storing to Database and Wrappers
Added wrappers for incidents and accelerometer readings. Also added code in ShakeDetector.java and BEMShakeNotifier.m to begin storing to database.

felipefelixarias committed:
First working version of shake detect.
Things to implement:
User-specified shakes
Binary classification of incidents using SVM
Integrating mode of transportation for iOS

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,251 @@
+package edu.berkeley.eecs.emission.cordova.shakedetect;
```
I think this filename needs to be `ShakeDetector.java` to match the classname.

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,251 @@
+                if ((xRotation > 5f) && (yRotation < 5f) && (zRotation) < 5f) {
+                    lastUpdated = currTime;
+                    Log.i(this, GyroTag, "," + Float.toString(xRotation) + "," );
+                    notifyEvent(this, POTENTIAL_INCIDENT, null);
```
This needs to be `storeDataAndNotify`, similar to the accelerometer code

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,251 @@
+            incidentData.put("inc_ts", ((double) currTime) / 1000);
+        } else {
+            SimpleLocation mostRecentLocation = loc[0];
+            incidentData.put("inc_ts", ((double) currTime) / 1000);
```
Instead of copying every single entry over, you should have this be a field `"inc"` whose value is the JSON representation of the SimpleLocation, as serialized by Gson (e.g. `new Gson().toJson(cfg))`). See https://github.com/google/gson for more details.

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,251 @@
+        */
+
+        incidentData.put("inc_xAccel", accelerations[0]);
+        incidentData.put("inc_yAccel", accelerations[1]);
```
Again, there should be a wrapper for the acceleration, and it should be stored as an object directly. That will make this much more modular.

shankari commented on src/android/shakeNotificationReceiver.java:
```diff
@@ -0,0 +1,81 @@
+
+    protected void manageShakeDetection( final Context context, final String eventName){
+
+        if (eventName.equals(context.getString(R.string.transition_exited_geofence))) {
```
Should be able to listen to the cleaned notifications (e.g. `TRIP_STARTED` and `TRIP_ENDED`, from 
https://github.com/e-mission/e-mission-transition-notify/blob/master/src/android/TransitionNotificationReceiver.java#L103) instead.

shankari replied:
Actually, I take this back. Using the cleaned notifications as-is will introduce a dependency between the transition notify plugin and this plugin, and dependencies are bad. Let's keep this using the raw transitions now. Eventually, I should generate the cleaned transitions right from the transition FSM.

shankari commented on src/android/shakeNotificationReceiver.java:
```diff
@@ -0,0 +1,81 @@
+            Intent intent = new Intent(context, ShakeDetector.class);
+            context.startService(intent);
+        }
+        else if (eventName.equals(context.getString(R.string.transition_stopped_moving)) || eventName.equals("local.transition.stopped_moving") || eventName.equals("local.transition.tracking_error")) {
```
then you don't need this more complex change.

shankari commented on src/android/shakeNotifier.java:
```diff
@@ -0,0 +1,250 @@
+            return true;
+        } 
+
+            final JSONObject localNotifyConfig = args.getJSONObject(1);
```
why do you have this second copy of the code?

shankari reviewed (changes requested):
Code will work, but should be cleaner and better engineered

shankari commented on src/ios/BEMShakeNotifier.m:
```diff
@@ -0,0 +1,431 @@
+    // copied over.
+    NSLog(@"BEMShakeNotifier:pluginInitialize singleton -> initialize statemachine and delegate");
+    __typeof(self) __weak weakSelf = self;
+    [[NSNotificationCenter defaultCenter] addObserverForName:CFCTransitionNotificationName object:nil queue:nil
```
why do you need this? iOS piggybacks on the location changes, so will only sense when the location tracking is on.

shankari commented on src/ios/BEMShakeNotifier.m:
```diff
@@ -0,0 +1,431 @@
+    motionManager = [[CMMotionManager alloc] init];
+    if([motionManager isAccelerometerAvailable]){
+        if([motionManager isAccelerometerActive] == NO){
+            [LocalNotificationManager addNotification:[NSString stringWithFormat:
```
Don't you need to actually process the accelerometer inputs before you know whether there is a shake or not?

shankari commented on src/ios/BEMShakeNotifier.m:
```diff
@@ -0,0 +1,431 @@
+            __block double lastUpdated = 0.0;
+            
+            double yAccelerationConst = 0.4;
+            [motionManager startAccelerometerUpdatesToQueue:[NSOperationQueue mainQueue] withHandler:^(CMAccelerometerData *data, NSError *error){
```
Does this actually collect data while in the background?

shankari replied:
apparently it does so if your app is already launched for other background operations.
https://stackoverflow.com/questions/21947864/getting-accelerometer-data-in-background-using-coremotion
https://stackoverflow.com/questions/8716466/how-nike-gps-on-iphone-receives-accelerometer-updates-in-the-background
For some time.
https://forums.developer.apple.com/thread/65074

shankari commented on src/ios/BEMShakeNotifier.m:
```diff
@@ -0,0 +1,431 @@
+                        else{
+                            //store time and other data on location
+                            
+                            SimpleLocation *lastLoc = ((SimpleLocation*)lastLocation.firstObject);
```
again, `SimpleLocation` should be serialized directly using `DataUtils`.

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,251 @@
+        return -1;
+    }
+
+    private void storeDataAndNotify(Context context, float[] accelerations, long currTime, float combinedAccel) throws JSONException{
```
Although the name of the method is `storeDataAndNotify`, it doesn't appear to actually store any data. But I though that @ariasfelipe saved to the usercache because he used the data to get initial results....

shankari replied:
I guess he just used the logs, not the database...

shankari commented on src/android/shakeDetector.java:
```diff
@@ -0,0 +1,251 @@
+                //Used to diminish the effect of the acceleration in the y-direction
+                float yAccelerationConst = 0.4f;
+
+                lastAcceleration = currAcceleration;
```
This is wrong. What happens if the service is restarted? the data needs to be stored to the usercache and retrieved from there.

shankari replied:
This only works if the app is running in the foreground or on a non-memory constrained device. And it doesn't store any data to be pushed to the server!

shankari commented on src/ios/BEMShakeNotifier.m:
```diff
@@ -0,0 +1,431 @@
+- (void)fireGenericTransitionFor:(NSString*) transition withUserInfo:(NSDictionary*) userInfo {
+    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Received platform-specific notification %@", transition] showUI:FALSE];
+    
+    if ([TripDiaryStateMachine instance].currState == kWaitingForTripStartState &&
```
again, this should use generic transitions

shankari commented on src/ios/BEMShakeNotifier.m:
```diff
@@ -0,0 +1,431 @@
+                double currTime = [BuiltinUserCache getCurrentTimeSecs];;
+                if((currTime - lastUpdated) >= 2.0){
+                    double lastAcceleration = currentAcceleration;
+                    currentAcceleration = sqrtf(data.acceleration.x*data.acceleration.x + data.acceleration.y*data.acceleration.y*yAccelerationConst*yAccelerationConst);
```
again, this needs to be stored into the database.

shankari reviewed (changes requested):
Main missing functionality is the storing to the database - otherwise everything looks like it should work!
